cirq/contrib/paulistring/recombine.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2018 The ops Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union, cast
+
+from cirq import ops, circuits
+
+from cirq.contrib.paulistring.pauli_string_dag import (
+    pauli_string_dag_from_circuit)
+
+
+def move_non_clifford_into_clifford(circuit_left: Union[circuits.Circuit,
+                                                        circuits.CircuitDag],
+                                    circuit_right: circuits.Circuit
+                                    ) -> circuits.Circuit:
+    if isinstance(circuit_left, circuits.CircuitDag):
+        string_dag = cast(circuits.CircuitDag, circuit_left)
+    else:
+        string_dag = pauli_string_dag_from_circuit(
+                        cast(circuits.Circuit, circuit_left))
+    output_ops = list(circuit_right.all_operations())
+
+    rightmost_nodes = (set(string_dag.nodes)
+                       - set(before for before, _ in string_dag.edges))
+
+    while rightmost_nodes:
+        # Pick the Pauli string that can be moved furthest through the Clifford
+        # circuit
+        shortest_string_len = float('inf')
+        best_index = 0
+        best_string_op = None
+        best_node = None
+
+        for right_node in rightmost_nodes:
```

Comment:
Take this loop, up to the `break` in the innermost loop, and put it integer a generator function that yields the values of `string_op`, `i`, and `right_node` iterated over. Then use

```python
best_string_op, best_index, best_node = max(
    extracted_function(...),
    key=lambda e: (len(e[0].pauli_string), e[1]))
```

Reply:
Done.